src/model/operation/batchnorm.cc

```diff
@@ -17,8 +17,187 @@ BatchNormHandle::BatchNormHandle(const float momentum, const Tensor& input) {
   } else {
     LOG(FATAL) << "The dimension of input should either be 4D or 2D.";
   }
+
+
+#ifdef USE_MKLDNN
+  dtype = GetMKLDNNDataType(input.data_type());
+  epsilon =1e-5f;
+  data_memory_format = is_2d ? mkldnn::memory::format::nc : mkldnn::memory::format::nchw;
+  if (is_2d) {
+    x_dims = {(int)batchsize, (int)channels};
+    y_dims = {(int)batchsize, (int)channels};
+  } else {
+    x_dims = {(int)batchsize, (int)channels, (int)height, (int)width};
+    y_dims = {(int)batchsize, (int)channels, (int)height, (int)width};
+  }
+
+  auto eng = *input.device()->context(0)->engine;
+  x_md = new mkldnn::memory::desc(x_dims, dtype, data_memory_format);
+  dx_md = new mkldnn::memory::desc(x_dims, dtype, data_memory_format);
+  bn_fwd_d = new mkldnn::batch_normalization_forward::desc(mkldnn::forward_training, *x_md, epsilon,
+                                                           mkldnn::use_scale_shift);
+  bn_fwd_pd = new mkldnn::batch_normalization_forward::primitive_desc(*bn_fwd_d, eng);
+#endif // USE_MKLDNN
+
 };
 
+
+  BatchNormHandle::~BatchNormHandle() {
+#ifdef USE_MKLDNN
+    delete (x_md);
+    delete (dx_md);
+    delete (bn_fwd_d);
+    delete (bn_fwd_pd);
+#endif // USE_MKLDNN
+  }
+
+#ifdef USE_MKLDNN
+
+  Tensor CpuBatchNormForwardInference(const BatchNormHandle &bnh, const Tensor& x, const Tensor& bnScale, const Tensor& bnBias,
+                                                        Tensor& running_mean, Tensor& running_var){
+
+    CHECK_EQ(x.device()->lang(), kCpp);
+    Tensor y;
+    y.ResetLike(x);
+
+
+    Tensor w = get_bn_weight_from(bnScale, bnBias);
+
+    y.device()->Exec([&y, &x, &w, &bnh](Context *ctx) {
+      try {
+        auto eng = *ctx->engine;
+        using namespace mkldnn;
+        auto x_mem = memory({{{bnh.x_dims}, bnh.dtype, bnh.data_memory_format}, eng}, x.block()->mutable_data());
```

Comment:
can we use x.block()->data() instead of mutable_data()?
data() returns a const pointer, i.e., the pointed memory can NOT be updated.
mutable_data() returns a raw pointer, i.e., the pointed memory can be updated.

Reply:
to feed the data into mkldnn, `memory primitive` is required. and constructor only accept `void * data_handle`. and it seems that this is the only way for now. https://github.com/intel/mkl-dnn/issues/387

Reply:
then keep mutable_data(). copying the data is expensive.